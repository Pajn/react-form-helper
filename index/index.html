<!DOCTYPE html><html><head><meta charSet="utf-8" class="next-head"/><link rel="preload" href="/react-form-helper/_next/static/P9qmuBmvlTecm1q8kpD8u/pages/index.js" as="script"/><link rel="preload" href="/react-form-helper/_next/static/P9qmuBmvlTecm1q8kpD8u/pages/_app.js" as="script"/><link rel="preload" href="/react-form-helper/_next/static/P9qmuBmvlTecm1q8kpD8u/pages/_error.js" as="script"/><link rel="preload" href="/react-form-helper/_next/static/runtime/webpack-42652fa8b82c329c0559.js" as="script"/><link rel="preload" href="/react-form-helper/_next/static/chunks/commons.397e029ed8da0a828c57.js" as="script"/><link rel="preload" href="/react-form-helper/_next/static/runtime/main-80afdc98fe79889eb792.js" as="script"/></head><body><div id="__next"></div><script>__NEXT_DATA__ = {"props":{"pageProps":{}},"page":"/index","query":{},"buildId":"P9qmuBmvlTecm1q8kpD8u","assetPrefix":"/react-form-helper","nextExport":true};__NEXT_LOADED_PAGES__=[];__NEXT_REGISTER_PAGE=function(r,f){__NEXT_LOADED_PAGES__.push([r, f])}</script><script async="" id="__NEXT_PAGE__/index" src="/react-form-helper/_next/static/P9qmuBmvlTecm1q8kpD8u/pages/index.js"></script><script async="" id="__NEXT_PAGE__/_app" src="/react-form-helper/_next/static/P9qmuBmvlTecm1q8kpD8u/pages/_app.js"></script><script async="" id="__NEXT_PAGE__/_error" src="/react-form-helper/_next/static/P9qmuBmvlTecm1q8kpD8u/pages/_error.js"></script><script src="/react-form-helper/_next/static/runtime/webpack-42652fa8b82c329c0559.js" async=""></script><script src="/react-form-helper/_next/static/chunks/commons.397e029ed8da0a828c57.js" async=""></script><script src="/react-form-helper/_next/static/runtime/main-80afdc98fe79889eb792.js" async=""></script></body></html>