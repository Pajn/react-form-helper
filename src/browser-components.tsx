import React from 'react'
import {Omit} from 'recompose'
import {FieldComponentProps, SaveButtonComponentProps} from './index'

export const BrowserButton = ({
  loading: _,
  ...props
}: SaveButtonComponentProps<any>) => <button {...props} />

export type BrowserInputProps = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  'value' | 'onChange'
> &
  FieldComponentProps<string> & {
    divProps?: React.HTMLAttributes<HTMLDivElement>
    labelProps?: React.HTMLAttributes<HTMLLabelElement>
    labelSpanProps?: React.HTMLAttributes<HTMLSpanElement>
  }
export const BrowserInput = ({
  error: _,
  onChange,
  label,
  divProps,
  labelProps,
  labelSpanProps,
  value = '',
  ...props
}: BrowserInputProps) => (
  <div {...divProps}>
    {label ? (
      <label {...labelProps}>
        <span {...labelSpanProps}>{label}</span>
        <input
          {...props}
          value={value}
          onChange={e =>
            onChange((e.target as HTMLInputElement)[
              props.type === 'checkbox' ? 'checked' : 'value'
            ] as string)
          }
        />
      </label>
    ) : (
      <input
        {...props}
        value={value}
        onChange={e =>
          onChange((e.target as HTMLInputElement)[
            props.type === 'checkbox' ? 'checked' : 'value'
          ] as string)
        }
      />
    )}
  </div>
)
