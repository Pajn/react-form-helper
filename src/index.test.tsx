import React from 'react'
import renderer from 'react-test-renderer'
import {BrowserInput, BrowserInputProps} from './browser-components'
import {FieldComponentProps, FormHelper, field, fields} from './index'

describe('FormHelper', () => {
  it('should render the passed fields', () => {
    expect(
      renderer.create(
        <FormHelper
          value={{a: 'a'}}
          onSave={() => {}}
          fields={[{path: ['a']}, {path: ['b']}]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should render using specified components', () => {
    expect(
      renderer.create(
        <FormHelper
          inputComponent="some-input"
          value={{a: 'a', c: 'c'}}
          onSave={() => {}}
          fields={[
            {path: ['a']},
            {path: ['b']},
            {component: 'other-input', path: ['c']},
            {component: 'other-input', path: ['d']},
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should not pass errors to the browser inputs', () => {
    expect(
      renderer.create(
        <FormHelper
          value={{}}
          onSave={() => {}}
          fields={[
            {path: ['a'], error: 'a error'},
            {
              path: ['b'],
              validationError: 'b',
              validations: {b: {text: 'b error'}},
            },
            {
              path: ['c'],
              validations: {c: {text: 'c error', test: () => false}},
            },
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should pass errors to custom inputs', () => {
    expect(
      renderer.create(
        <FormHelper
          inputComponent="input"
          value={{c: 'value'}}
          onSave={() => {}}
          fields={[
            {path: ['a'], error: 'a error'},
            {
              path: ['b'],
              validationError: 'b',
              validations: {b: {text: 'b error'}},
            },
            {
              path: ['c'],
              validations: {c: {text: 'c error', test: () => false}},
            },
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should not run custom validators on no value', () => {
    expect(
      renderer.create(
        <FormHelper
          inputComponent="input"
          value={{}}
          onSave={() => {}}
          fields={[
            {
              path: ['a'],
              validations: {
                a: {
                  text: 'a error',
                  test: () => {
                    throw 'should not be called'
                  },
                },
              },
            },
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should pass errors to untouched inputs if errorOnTouched is true', () => {
    expect(
      renderer.create(
        <FormHelper
          errorOnTouched
          inputComponent="input"
          value={{c: 'value'}}
          onSave={() => {}}
          fields={[
            {path: ['a'], error: 'a error'},
            {
              path: ['b'],
              validationError: 'b',
              validations: {b: {text: 'b error'}},
            },
            {
              path: ['c'],
              validations: {c: {text: 'c error', test: () => false}},
            },
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should pass errors to untouched inputs if they are listed in the errorOnTouched array', () => {
    expect(
      renderer.create(
        <FormHelper
          errorOnTouched={['b']}
          inputComponent="input"
          value={{c: 'value'}}
          onSave={() => {}}
          fields={[
            {path: ['a'], error: 'a error'},
            {
              path: ['b'],
              validationError: 'b',
              validations: {b: {text: 'b error'}},
            },
            {
              path: ['c'],
              validations: {c: {text: 'c error', test: () => false}},
            },
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should pass additional props down to the field components', () => {
    expect(
      renderer.create(
        <FormHelper
          value={{}}
          onSave={() => {}}
          fields={[{path: ['a'], a: 'A', props: {b: 'B'}}]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should support rendering fields based on value', () => {
    expect(
      renderer.create(
        <FormHelper
          value={{a: 'a'}}
          onSave={() => {}}
          fields={[
            value => (value.a === 'a' ? [{path: ['a']}, {path: ['b']}] : []),
            value => (value.b === 'b' ? [{path: ['c']}] : []),
          ]}
        />,
      ),
    ).toMatchSnapshot()
    expect(
      renderer.create(
        <FormHelper
          value={{a: 'a', b: 'b'}}
          onSave={() => {}}
          fields={[
            value => (value.a === 'a' ? [{path: ['a']}, {path: ['b']}] : []),
            value => (value.b === 'b' ? [{path: ['c']}] : []),
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should support mapping rendered fields', () => {
    expect(
      renderer.create(
        <FormHelper
          inputComponent="some-input"
          value={{a: 'a', c: 'c'}}
          onSave={() => {}}
          fields={[
            {path: ['a']},
            {path: ['b']},
            {component: 'other-input', path: ['c']},
            {component: 'other-input', path: ['d']},
          ]}
          mapField={field => <div>{field}</div>}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('should pass disabled to the fields', () => {
    expect(
      renderer.create(
        <FormHelper
          value={{a: 'a'}}
          onSave={() => {}}
          fields={[{path: ['a'], disabled: true}, {path: ['b']}]}
        />,
      ),
    ).toMatchSnapshot()

    expect(
      renderer.create(
        <FormHelper
          value={{a: 'a'}}
          onSave={() => {}}
          fields={[
            {path: ['a'], disabled: object => object.a === 'a'},
            {path: ['b'], disabled: object => object.a === 'b'},
          ]}
        />,
      ),
    ).toMatchSnapshot()
  })

  describe('typing helpers', () => {
    type User = {id: string; name: string; birthdate?: Date}

    it('should pass through values', () => {
      expect(
        fields<User>([
          field<User, BrowserInputProps, string>(BrowserInput, {
            path: ['name'],
            label: 'Name',
            onChange: a => {},
            props: {name: 'name'},
          }),
        ]),
      ).toMatchSnapshot()
    })
  })

  describe('dirty check', () => {
    it('should disable the save button if form has not changed', () => {
      expect(
        renderer.create(
          <FormHelper
            value={{a: 'a'}}
            onSave={() => {}}
            fields={[
              {path: ['a'], disabled: object => object.a === 'a'},
              {path: ['b'], disabled: object => object.a === 'b'},
            ]}
            dirtyCheck
            saveButton="save"
          />,
        ),
      ).toMatchSnapshot()
    })

    it('should enable the save button when the form has changed', () => {
      const subject = renderer.create(
        <FormHelper
          value={{a: 'a'}}
          onSave={() => {}}
          fields={[
            {path: ['a'], disabled: object => object.a === 'a'},
            {path: ['b'], disabled: object => object.a === 'b'},
          ]}
          dirtyCheck
          saveButton="save"
        />,
      )
      subject.root
        .findAllByType('input')[0]
        .props.onChange({target: {value: 'c'}})

      expect(subject).toMatchSnapshot()
    })
  })
})
