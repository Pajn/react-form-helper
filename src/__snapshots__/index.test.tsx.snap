// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHelper dirty check should disable the save button if form has not changed 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      disabled={true}
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <input
      disabled={false}
      onChange={[Function]}
      value=""
    />
  </div>
  <button
    disabled={true}
    type="submit"
    value={
      Object {
        "a": "a",
      }
    }
  >
    save
  </button>
</form>
`;

exports[`FormHelper dirty check should enable the save button when the form has changed 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      disabled={false}
      onChange={[Function]}
      value="c"
    />
  </div>
  <div>
    <input
      disabled={false}
      onChange={[Function]}
      value=""
    />
  </div>
  <button
    disabled={false}
    type="submit"
    value={
      Object {
        "a": "c",
      }
    }
  >
    save
  </button>
</form>
`;

exports[`FormHelper should not pass errors to the browser inputs 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper should not run custom validators on no value 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
  />
</form>
`;

exports[`FormHelper should pass additional props down to the field components 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      a="A"
      b="B"
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper should pass disabled to the fields 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      disabled={true}
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper should pass disabled to the fields 2`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      disabled={true}
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <input
      disabled={false}
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper should pass errors to custom inputs 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    error="a error"
    onChange={[Function]}
  />
  <input
    error="b error"
    onChange={[Function]}
  />
  <input
    error="c error"
    onChange={[Function]}
    value="value"
  />
</form>
`;

exports[`FormHelper should pass errors to untouched inputs if errorOnTouched is true 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onBlur={[Function]}
    onChange={[Function]}
  />
  <input
    onBlur={[Function]}
    onChange={[Function]}
  />
  <input
    onBlur={[Function]}
    onChange={[Function]}
    value="value"
  />
</form>
`;

exports[`FormHelper should pass errors to untouched inputs if they are listed in the errorOnTouched array 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    error="a error"
    onChange={[Function]}
  />
  <input
    onBlur={[Function]}
    onChange={[Function]}
  />
  <input
    error="c error"
    onChange={[Function]}
    value="value"
  />
</form>
`;

exports[`FormHelper should render the passed fields 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper should render using specified components 1`] = `
<form
  onSubmit={[Function]}
>
  <textarea
    onChange={[Function]}
    value="a"
  />
  <textarea
    onChange={[Function]}
  />
  <other-input
    onChange={[Function]}
    value="c"
  />
  <other-input
    onChange={[Function]}
  />
</form>
`;

exports[`FormHelper should support mapping rendered fields 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <textarea
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <textarea
      onChange={[Function]}
    />
  </div>
  <div>
    <other-input
      onChange={[Function]}
      value="c"
    />
  </div>
  <div>
    <other-input
      onChange={[Function]}
    />
  </div>
</form>
`;

exports[`FormHelper should support rendering fields based on value 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper should support rendering fields based on value 2`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <input
      onChange={[Function]}
      value="a"
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value="b"
    />
  </div>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
  </div>
</form>
`;

exports[`FormHelper typing helpers should pass through values 1`] = `
Array [
  Object {
    "component": [Function],
    "label": "Name",
    "onChange": [Function],
    "path": Array [
      "name",
    ],
    "props": Object {
      "name": "name",
    },
  },
]
`;
